### PR TITLE
Use `include-what-you-use` on lib/core

### PR DESCRIPTION
**Pull request by andy31415**
### Changes:

ran:

```
iwyu_tool.py \
    -p out/linux-x64-all-clusters-clang/compile_commands.json  \
    src/lib/core/ \
    -- \
    -Xiwyu --no_comments \
    -Xiwyu --comment_style=none \
    -Xiwyu --cxx17ns  \
    -Xiwyu no_fwd_decls \
  | tee out/iwyu.out

cd out/linux-x64-all-clusters-clang

fix_includes.py <../iwyu.out 
```

After this I had to make some fixes:
  - IWYU for some reason added some extra forward declares which I removed (this may need eventual fixing...)
  - fixed SystemBuildConfig and CHIPBuildConfig includes since those are private. Moved them to SystemConfig/CHIPConfig instead
  - re-arranged some of the includes to split system and chip includes as well as removed some redundant `@file` comments.
  
### Testing

Compile should succeed.

### CODE REVIEW

**Review by arkq — APPROVED** [review]

File: src/lib/core/TLVReader.cpp [hunk 1]
```diff
@@ -15,24 +15,32 @@
...
  */
+#include <lib/core/TLVReader.h>
 
-/**
- *    @file
- *      This file implements a parser for the CHIP TLV (Tag-Length-Value) encoding format.
- *
- */
+#include <stdint.h>
+#include <string.h>
```
Comment (tehampson) [inline src/lib/core/TLVReader.cpp:21]: Is there a reason why sometimes included the standard lib headers is above SDK headers and sometimes it is not? I would like to know why it doesn't appear to be consistent

Comment (andy31415) [reply src/lib/core/TLVReader.cpp:21]: No real reason - I just copied and pasted what was already there without trying to re-order. I belive splitting them from the rest with newlines makes sense. However where they should go I am unclear about.

If forced to make a choice I would put them last to limit chances of accidental dependency on include order, however not sure if I should try to update the PR for that as the updates were mostly automatic with me only adding some whitespace.

Comment (arkq) [reply src/lib/core/TLVReader.cpp:21]: When I see such inconsistent I'm trying to follow google C++ styleguide: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

**Review by tehampson — APPROVED** [review]

File: src/lib/core/TLVReader.cpp (hunk 1 above)
Comment (andy31415) [reply src/lib/core/TLVReader.cpp:21]: +1

I will try to update includes as I go by if changing individual files. For large changes for now though, I would prefer to limit change amount.

Comment (tehampson) [reply src/lib/core/TLVReader.cpp:21]: I guess if we want to standardize this we should do so through automation. I am a little surprised some formatter/linter wasn't already enforcing this.

What you have done is much better than what was there before and doesn't create an issue. Feel free to mark this as resolved

**Pull request merged**